crates/ruff_cli/src/cache.rs

```diff
@@ -1,171 +1,235 @@
+        let writer = BufWriter::new(file);
+        bincode::serialize_into(writer, &self).with_context(|| {
+            format!(
+                "Failed to serialise cache to file '{}'",
+                self.path.display()
+            )
+        })
+    }
 
-        for message in self.messages {
-            // Using a Vec instead of a HashMap because the cache is per file and the large majority of
-            // files have exactly one source file.
-            let file_id = if let Some(position) = files
-                .iter()
-                .position(|(filename, _)| *filename == message.filename())
-            {
-                position
-            } else {
-                let index = files.len();
-                files.push((message.filename(), message.file.source_text()));
-                index
-            };
+    /// Returns the relative path based on `path` and the package root.
+    ///
+    /// Returns `None` if `path` is not within the package.
+    pub(crate) fn relative_path<'a>(&self, path: &'a Path) -> Option<&'a RelativePath> {
+        path.strip_prefix(&self.package_root).ok()
+    }
 
-            s.serialize_element(&SerializeMessage { message, file_id })?;
+    /// Get the cached results for a single file at relative `path`. This uses
+    /// `file_last_modified` to determine if the results are still accurate
+    /// (i.e. if the file hasn't been modified since the cached run).
+    ///
+    /// This returns `None` if `file_last_modified` differs from the cached
+    /// timestamp or if the cache doesn't contain results for the file.
+    pub(crate) fn get(
+        &self,
+        path: &RelativePath,
+        file_last_modified: SystemTime,
+    ) -> Option<FileCache> {
+        let file = self.files.lock().unwrap().get(path)?.clone();
+
+        // Make sure the file hasn't changed since the cached run.
+        if file.last_modified != file_last_modified {
+            return None;
         }
 
-        s.end()
+        Some(file)
+    }
+
+    /// Add or update a file cache at `path` relative to the package root.
+    pub(crate) fn update(&self, path: RelativePathBuf, file: FileCache) {
+        self.files.lock().unwrap().insert(path, file);
     }
-}
 
-struct SerializeMessage<'a> {
-    message: &'a Message,
-    file_id: usize,
+    /// Remove a file cache at `path` relative to the package root.
+    pub(crate) fn remove(&self, path: &RelativePath) {
+        self.files.lock().unwrap().remove(path);
+    }
 }
 
-impl Serialize for SerializeMessage<'_> {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        let Message {
-            kind,
-            range,
-            fix,
-            // Serialized manually for all files
-            file: _,
-            noqa_offset: noqa_row,
-        } = self.message;
+/// [`Path`] that is relative to the package root in [`PackageCache`].
+pub(crate) type RelativePath = Path;
```

Comment:
That's helpful, i've already been wondering about this when reading the code above